src/app/rpmostree-compose-builtin-tree.c

```diff
@@ -800,6 +805,130 @@ process_touch_if_changed (GError **error)
+     mount --make-private /mnt/mock-mount
+     mount --bind /mnt/mock-mount /mnt/mock-mount
+     for mnt in proc sys; do
+     mount --bind /$mnt /mnt/mock-mount/$mnt
+     done
+
+     chroot /mnt/mock-mount "$@"
+    */
+  if (unshare (CLONE_NEWNS) < 0)
+    return glnx_throw_errno_prefix (error, "unshare(CLONE_NEWNS)");
+  const char mock_mntdir[] = "/mnt/mock-mount";
+  if (!glnx_shutil_mkdir_p_at (AT_FDCWD, mock_mntdir, 0755, NULL, error))
+    return FALSE;
+  if (mount ("/", mock_mntdir, NULL, MS_BIND | MS_PRIVATE, NULL) < 0)
+    return glnx_throw_errno_prefix (error, "bindmount(/)");
+  if (mount (mock_mntdir, mock_mntdir, NULL, MS_BIND, NULL) < 0)
+    return glnx_throw_errno_prefix (error, "bindmount(%s)", mock_mntdir);
+  const char *subdirs[] = { "proc", "sys", "srv" };
+  for (guint i = 0; i < G_N_ELEMENTS (subdirs); i++)
+    {
+      g_autofree char *src_subpath = g_strconcat ("/", subdirs[i], NULL);
+      g_autofree char *dest_subpath = g_strconcat (mock_mntdir, "/", subdirs[i], NULL);
+      if (mount (src_subpath, dest_subpath, NULL, MS_BIND, NULL) < 0)
+        return glnx_throw_errno_prefix (error, "bindmount(%s)", src_subpath);
+    }
+
+  if (chdir (mock_mntdir) < 0)
+    return glnx_throw_errno_prefix (error, "chdir(%s)", mock_mntdir);
+  if (chroot (mock_mntdir) < 0)
+    return glnx_throw_errno_prefix (error, "chroot");
+  if (chdir (orig_pwd) < 0)
+    return glnx_throw_errno_prefix (error, "chdir(%s)", orig_pwd);
+  /* See also the bits in bwrap that do this */
+  setenv ("PWD", orig_pwd, 1);
+
+  /* Avoid recursion */
+  setenv ("RPMOSTREE_MOCK_COMPAT_RECURSE", "1", TRUE);
+
+  if (execvp (arr->pdata[0], (char**)arr->pdata) < 0)
+    return glnx_throw_errno_prefix (error, "execvp");
+
+  g_assert_not_reached ();
+  return TRUE;
+}
+
+static gboolean
+run_bwrap_selftest (GError **error)
+{
+  /* Test whether or not bwrap is going to work - we will fail inside e.g. a Docker
+   * container without --privileged or userns exposed.  This code has also
+   * gained some gymnastics to handle being invoked inside `mock --old-chroot`.
+   */
+  const gboolean explicit_mock_compat_mode = opt_mock_compat || getenv ("RPMOSTREE_MOCK_COMPAT");
+  gboolean mock_compat_mode = explicit_mock_compat_mode;
+  if (!explicit_mock_compat_mode)
+    {
+      /* Does /builddir exist and it's owned by `mockbuild` ? */
+      struct stat stbuf;
+      if (!glnx_fstatat_allow_noent (AT_FDCWD, "/builddir", &stbuf, AT_SYMLINK_NOFOLLOW, error))
+        return FALSE;
+      if (errno == 0)
+        {
+          struct passwd *pwent = getpwnam ("mockbuild");
+          if (pwent)
+            mock_compat_mode = (pwent->pw_uid == stbuf.st_uid);
+        }
+    }
+
+  if (mock_compat_mode)
+    {
+      g_printerr ("NOTE: Mock compatibility mode %senabled\n",
+                  explicit_mock_compat_mode ? "explicitly " : "");
+      g_autoptr(GError) local_error = NULL;
+      if (!rpmostree_bwrap_selftest (&local_error))
```

Comment:
So even in the explicit `--mock-compat` mode, we might not actually do the re-exec then? Seems slightly counter-intuitive. I'd expect `--mock-compat` would always re-exec, and otherwise only re-exec if we auto-detect we're running in a mock container with classical chroot.

Reply:
Alternatively, maybe at least a follow-up notice to say we didn't *actually* apply a workaround even though `--mock-compat` was passed?

Reply:
I may just ditch the explicit `--mock-compat` arg since I think our automatic detection here should be pretty reliable - when else would one have a `/builddir` directory owned by `mockbuild`?  And further requiring a  new argument here is a versionmatrix hazard of course.

Reply:
Yeah, that works for me!